> Why? You can just do -fsys=gtk4-layer-shell project-wide to use the system library and avoid compiling yourself. 
This won't mean compiling gtk4-layer-shell in the system-integration, only using the system-installed headers instead of the ones downloaded via the zig build system.

Our existing system-integrated libraries already prefer using pkgconfig names in the b.systemIntegration(<name>) instead of shared-object names. By specifying a pkgconfig name the build system can use it to find system headers and resolve the linking tree. 
```
> pkgconf --cflags --shared --libs gtk4-layer-shell-0
-I/usr/include/gtk4-layer-shell -I/usr/include/gtk-4.0 -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/webp -I/usr/include/cairo -I/usr/include/libxml2 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/pixman-1 -I/usr/include/graphene-1.0 -I/usr/lib64/graphene-1.0/include -mfpmath=sse -msse -msse2 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -DWITH_GZFILEOP -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/sysprof-6 -pthread -lgtk4-layer-shell -lpangocairo-1.0 -lpango-1.0 -lgdk_pixbuf-2.0 -lcairo-gobject -lcairo -lvulkan -lgraphene-1.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0
``` 

Otherwise, I want to minimise the bundled files (instead of system packaged) needed to build ghostty. It seems the package reviewers for fedora linux at https://bugzilla.redhat.com/show_bug.cgi?id=2334903 prefer it. 
In packaging for official channels in Fedora Linux, I have to audit the bundled dependencies and track down the licenses for all source files that are bundled into a `ghostty.src.rpm` which can perform a standalone offline build which specifies system-packages installed before the offline phase and contains bundled archives of other downloads. Minimising the size of this `.src.rpm` is preferred for packaging.
```sh
$ zig build --fetch --global-cache-dir $(pwd)/cache-thing
// Remove contents of the downloaded gtk4-layer-shell (including header files)
$ rm -r $(pwd)/cache-thing/p/12203eff4829ad8afdd828eb323d48e5ba8dbb44d224e9e314d4ab1533c2bec20f4b/*
$ zig build --system --global-cache-dir $(pwd)/cache-thing/p --help
// gtk4-layer-shell is system-integrated
$ zig build --system --global-cache-dir $(pwd)/cache-thing/p
```
On main this process errors as we rely on downloading the headers to dynamic link against instead of using the system-installed headers + system-installed compiled shared-lib.
```
install
└─ install ghostty
   └─ zig build-exe ghostty Debug native 2 errors
pkg/gtk4-layer-shell/src/main.zig:1:11: error: C import failed
const c = @cImport({
          ^~~~~~~~
referenced by:
    ShellLayer: pkg/gtk4-layer-shell/src/main.zig:7:18
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
/home/fedora/dev/ghostty/.zig-cache/o/01084885f82067596db10f997758cb59/cimport.h:1:10: error: 'gtk4-layer-shell.h' file not found
#include <gtk4-layer-shell.h>
```
> How would the system integration option propagate from the main project to the gtk4-layer-shell lib?
The sub-dependency is passed the Build object when declared as a dependency, `b.dependency(module: "gtk4_layer_shell")` in ghostty's `build.zig` passes the Build into the `build.zig.zon` dep with the module name and includes the module as a build-step in ghostty's build. This also lets it inherit ghostty's build and system-integration options, though technically it can override stuff there.

I think we've also kept our `pkg/` as able to be compiled standalone and taken as dependencies for other projects, perhaps in the future we'll be able to split them as separate zig-packages.

Not sure if it's a real problem, but we could potentially have mismatches between our downloaded headers and the system shared-object.

--- a/pkg/gtk4-layer-shell/build.zig
+++ b/pkg/gtk4-layer-shell/build.zig
@@ -3,12 +3,14 @@ const std = @import("std");
 // TODO: Import this from build.zig.zon when possible
 const version: std.SemanticVersion = .{ .major = 1, .minor = 1, .patch = 0 };
 
+const dynamic_link_opts: std.Build.Module.LinkSystemLibraryOptions = .{
+    .preferred_link_mode = .dynamic,
+    .search_strategy = .mode_first,
+};
+
 pub fn build(b: *std.Build) !void {
     const target = b.standardTargetOptions(.{});
     const optimize = b.standardOptimizeOption(.{});
-
-    const upstream = b.dependency("gtk4_layer_shell", .{});
-    const wayland_protocols = b.dependency("wayland_protocols", .{});
 
     // Zig API
     const module = b.addModule("gtk4-layer-shell", .{
@@ -16,10 +18,25 @@ pub fn build(b: *std.Build) !void {
         .target = target,
         .optimize = optimize,
     });
-    module.addIncludePath(upstream.path("include"));
     // Needs the gtk.h header
     module.linkSystemLibrary("gtk4", dynamic_link_opts);
 
+    if (b.systemIntegrationOption("gtk4-layer-shell", .{})) {
+        module.linkSystemLibrary("gtk4-layer-shell-0", dynamic_link_opts);
+    } else {
+        _ = try buildLib(b, module, .{
+            .target = target,
+            .optimize = optimize,
+        });
+    }
+}
+
+fn buildLib(b: *std.Build, module: *std.Build.Module, options: anytype) !*std.Build.Step.Compile {
+    const target = options.target;
+    const optimize = options.optimize;
+
+    const upstream = b.dependency("gtk4_layer_shell", .{});
+    const wayland_protocols = b.dependency("wayland_protocols", .{});
     // Shared library
     const lib = b.addSharedLibrary(.{
         .name = "gtk4-layer-shell",
@@ -29,6 +46,7 @@ pub fn build(b: *std.Build) !void {
     lib.linkLibC();
     lib.addIncludePath(upstream.path("include"));
     lib.addIncludePath(upstream.path("src"));
+    module.addIncludePath(upstream.path("include"));
 
     // GTK
     lib.linkSystemLibrary2("gtk4", dynamic_link_opts);
@@ -76,6 +94,16 @@ pub fn build(b: *std.Build) !void {
         .{ .include_extensions = &.{".h"} },
     );
 
+    // Certain files relating to session lock were removed as we don't use them
+    const srcs: []const []const u8 = &.{
+        "gtk4-layer-shell.c",
+        "layer-surface.c",
+        "libwayland-shim.c",
+        "registry.c",
+        "stolen-from-libwayland.c",
+        "stubbed-surface.c",
+        "xdg-surface-server.c",
+    };
     lib.addCSourceFiles(.{
         .root = upstream.path("src"),
         .files = srcs,
@@ -87,20 +115,5 @@ pub fn build(b: *std.Build) !void {
     });
 
     b.installArtifact(lib);
+    return lib;
 }
-
-// Certain files relating to session lock were removed as we don't use them
-const srcs: []const []const u8 = &.{
-    "gtk4-layer-shell.c",
-    "layer-surface.c",
-    "libwayland-shim.c",
-    "registry.c",
-    "stolen-from-libwayland.c",
-    "stubbed-surface.c",
-    "xdg-surface-server.c",
-};
-
-const dynamic_link_opts: std.Build.Module.LinkSystemLibraryOptions = .{
-    .preferred_link_mode = .dynamic,
-    .search_strategy = .mode_first,
-};

--- a/src/build/SharedDeps.zig
+++ b/src/build/SharedDeps.zig
@@ -506,7 +506,7 @@ pub fn add(
                     // IMPORTANT: gtk4-layer-shell must be linked BEFORE
                     // wayland-client, as it relies on shimming libwayland's APIs.
                     if (b.systemIntegrationOption("gtk4-layer-shell", .{})) {
-                        step.linkSystemLibrary2("gtk4-layer-shell", dynamic_link_opts);
+                        step.linkSystemLibrary2("gtk4-layer-shell-0", dynamic_link_opts);
                     } else {
                         // gtk4-layer-shell *must* be dynamically linked,
                         // so we don't add it as a static library

